ENH: Restore original convert_objects and add _convert

bashtage opened the pull request:
Restores the v0.16 behavior of convert_objects and moves the new
version of _convert
Adds to_numeric for directly converting numeric data

closes #11116
closes #11133

Commit:
ENH: Resotre original convert_objects and add _convert

Restores the v0.16 behavior of convert_objects and moves the new
version of _convert
Adds to_numeric for directly converting numeric data

closes #11116
closes #11133

bashtage commented:
Still a few things to do:

 [] Update release notes to remove new docs and add deprecation note
 [] Add note about bugs fixed in `convert_objects`
 [] Restore docs for `convert_objects`
 [] Add tests for `to_numeric`
 [] Add docs for `to_numeric`

jreback commented on pandas/core/common.py:
```diff
@@ -1857,23 +1857,84 @@ def _maybe_box_datetimelike(value):
-                              timedelta=True,
-                              coerce=False,
+# TODO: Remove in 0.18 or 2017, which ever is sooner
+def _possibly_convert_objects(values, convert_dates=True,
```
How about we put ``_possibly_convert_objects`` and ``_soft_convert_objects`` in ``pandas/tools/convert.py`` (new file)
``common.py`` is just so huge. We can eventually move more stuff their as well. It will have to import ``common`` for now.

jreback replied:
I can do this after as well.

shoyer replied:
I'll just note that there is some value in keeping this function intact in `core._common`. Xray was using this private function internally, with a broken guard against it changing (see xray/xray#569 for details). Obviously, that was our fault and we've fixed it now.

jreback replied:
@shoyer I don't think you should be using this internal function (I saw you removed it). and instead use the public ``.convert_objects``

shoyer replied:
Fair enough -- but there's not any other way to hook directly into pandas' object array of datetimes -> datetime64 conversion. Might be worth thinking about making such an API at some point.

On Wed, Sep 23, 2015 at 5:58 PM, Jeff Reback <notifications@github.com>
wrote:

>> @@ -1857,23 +1857,84 @@ def _maybe_box_datetimelike(value):
>>  
>>  _values_from_object = lib.values_from_object
>>  
>> -
>> -def _possibly_convert_objects(values,
>> -                              datetime=True,
>> -                              numeric=True,
>> -                              timedelta=True,
>> -                              coerce=False,
>> +# TODO: Remove in 0.18 or 2017, which ever is sooner
>> +def _possibly_convert_objects(values, convert_dates=True,
> @shoyer I don't think you should be using this internal function (I saw you removed it). and instead use the public ``.convert_objects``
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/pydata/pandas/pull/11173/files#r40276398

jreback replied:
you mean like this?

```
In [20]: pd.to_datetime([datetime.datetime(2012,1,1,0,0),datetime.datetime(2013,1,1,1,1)],box=False)       
Out[20]: 
array(['2011-12-31T19:00:00.000000000-0500',
       '2012-12-31T20:01:00.000000000-0500'], dtype='datetime64[ns]')
```

shoyer replied:
Not quite -- I only want soft conversion (like the `pd.Series` constructor). I don't want to convert text that matches a regex, only an array of all `datetime` instances (to datetime64) or all `timedelta` instances (to timedelta64).

bashtage commented:
The important changes are:

`convert` in `ObjectManager`: https://github.com/pydata/pandas/pull/11173/files#diff-e705e723b2d6e7c0e2a0443f80916abfR1521

This change requires this to 'route' to either the old or new style converter. 

`to_numeric`: https://github.com/pydata/pandas/pull/11173/files#diff-40d1f19921aa9f2932951124fad07cc6R53

New function.

jorisvandenbossche commented:
@bashtage Really thanks a lot for this! 

(I am still not fully convinced the deprecation is needed, but I will try to summarize my reasons tomorrow)

jreback commented:
@bashtage 

- pls move the 2 converting functions either to a new module, ``core/convert.py``. 
- let's finish up the documentation changes.

pls let me know asap.

Commit:
REF: Move conversion functions to core/convert.py

Move conversion function to core/convert.py
Restore docs for convert_objects
Add bug fix notes for restores convert_objects
Ensure copy is passed to _possibly_convert_objects when needed

bashtage commented:
Still no tests for `to_numeric` or docs for this method.

jreback commented:
looks good! ping when tests for ``pd.to_numeric``

jreback commented:
@jorisvandenbossche ok with this

jorisvandenbossche commented:
`to_numeric` needs some fixes (and tests): `lib` is not yet imported, `lib.maybe_convert_numeric` expects an array so it bugs on a series as arg, and the final result needs to be boxed into a series again if the arg was a series.

bashtage commented:
Yes - it definitely needs work.  It expects something with the buffer
interface, so I need to ensure that this works as expected.


On Sun, Sep 27, 2015 at 8:12 PM Joris Van den Bossche <
notifications@github.com> wrote:

> to_numeric needs some fixes (and tests): lib is not yet imported,
> lib.maybe_convert_numeric expects an array so it bugs on a series as arg,
> and the final result needs to be boxed into a series again if the arg was a
> series.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pydata/pandas/pull/11173#issuecomment-143606211>.
>

jreback commented:
@bashtage did you have a chance to add some tests here?

Commit:
FIX: Add boxing for to_numeric and add tests

Added support for unboxing/boxing to to_numeric
Added tests for to_numeric

bashtage commented:
Should be ready when green I think

bashtage commented:
Can squash commits - wanted to make it easier to see changes at each stage since close to release

bashtage commented:
@jreback Green - let me know if I missed anything.  Only tick left is to add some docs although this can probably wait since `convert_objects` is still available.

jorisvandenbossche commented on pandas/tools/util.py:
```diff
@@ -48,3 +48,66 @@ def compose(*funcs):
     return reduce(_compose2, funcs)
+
+
+def to_numeric(arg, errors='raise', box=True, coerce=None):
```
'coerce' shouldn't be here?

jorisvandenbossche commented on pandas/tools/util.py:
```diff
@@ -48,3 +48,66 @@ def compose(*funcs):
+    arg : string, array of strings, mixed object array
+    errors : {'ignore', 'raise', 'coerce'}, default 'raise'
+        - If 'raise', then invalid parsing will raise an exception
+        - If 'coerce', then invalid parsing will be set as NaT
```
NaT -> NaN

jorisvandenbossche commented on pandas/tools/util.py:
```diff
@@ -48,3 +48,66 @@ def compose(*funcs):
+        - If 'ignore', then invalid parsing will return the input
+    box : boolean, default True
+        - If True returns a Series
+        - If False returns ndarray of values.
```
The `box` keyword here has another meaning as the `box` keyword in `to_datetime`. There, it is to determine of an index or an array is returned (a Series is always returned as a Series, regardless of the value of `box`).

But, the meaning of `to_datetime` is not very useful here I think. @bashtage Do you think this keyword is needed? (because, one option to not have this inconsistency, is maybe to just leave it out here? ) If you want the values itself as the return value, you can always pass it the values yourself instead of passing an argument (`pd.to_numeric(s.values)` instead of `pd.to_numeric(s, box=False)`

jreback replied:
I agree. let's remove ``box``.

jorisvandenbossche commented on pandas/tests/test_util.py:
```diff
@@ -76,6 +78,59 @@ def test_rands_array():
+
+        res = to_numeric(s, errors='coerce')
+        expected = pd.Series([1, -3.14, np.nan])
+        assert_series_equal(res, expected)
```
Can you add a test here with all NaNs? (as this was a behaviour in convert_objects that was changed)

eg

```
In [4]: s = pd.Series(['pear', 'orange', 'apple'])

In [5]: pd.to_numeric(s, errors='coerce')
Out[5]:
0   NaN
1   NaN
2   NaN
dtype: float64
```

jreback commented on doc/source/whatsnew/v0.17.0.txt:
```diff
@@ -992,6 +927,7 @@ Deprecations
 - ``Series.is_time_series`` deprecated in favor of ``Series.index.is_all_dates`` (:issue:`11135`)
 - Legacy offsets (like ``'A@JAN'``) listed in :ref:`here <timeseries.legacyaliases>` are deprecated (note that this has been alias since 0.8.0), (:issue:`10878`)
 - ``WidePanel`` deprecated in favor of ``Panel``, ``LongPanel`` in favor of ``DataFrame`` (note these have been aliases since < 0.11.0), (:issue:`10892`)
+- ``DataFrame.convert_objects`` has been deprecated in favor of type-specific function ``pd.to_datetime``, ``pd.to_timestamp`` and ``pd.to_numeric``.
```
add this issue number here

jreback commented on pandas/core/convert.py:
```diff
@@ -0,0 +1,132 @@
+        elif timedelta:
+            return pd.to_timedelta(values, errors='coerce', box=False)
+        elif numeric:
+            return lib.maybe_convert_numeric(values, set(), coerce_numeric=True)
```
an use ``pd.to_numeric`` here, no?

jreback commented on pandas/tests/test_series.py:
```diff
@@ -6449,21 +6449,122 @@ def test_apply_dont_convert_dtype(self):
     def test_convert_objects(self):
+
+        s = Series([1., 2, 3], index=['a', 'b', 'c'])
+        result = s.convert_objects(convert_dates=False, convert_numeric=True)
```
so anytime you use ``.convert_objects`` it needs a:

```
tm.assert_produces_warning(FutureWarning):
   .....
```

around, e.g. no need to show the deprecation warning.

jreback commented on pandas/tests/test_util.py:
```diff
@@ -76,6 +78,59 @@ def test_rands_array():
+        res = to_numeric(np.array(s), box=False, errors='ignore')
+        expected = np.array([1, -3.14, 'apple'], dtype='O')
+        for x,y in zip(res, expected):
+            assert_true(x == y)
```
use ``tm.assert_numpy_array_equal``

jreback commented on pandas/tools/tests/test_util.py:
```diff
@@ -89,6 +90,14 @@ def test_set_locale(self):
 
+class TestToNumeric(tm.TestCase):
+    def test_basic_conversion(self):
+        raw = pd.Series(['1','2.0',3])
```
move to the testing above (e.g. ``test_util``) with the other tests for ``to_numeric``

jreback commented on pandas/tools/util.py:
```diff
@@ -48,3 +48,66 @@ def compose(*funcs):
+    arg = com._ensure_object(arg)
+
+    coerce_numeric = False if errors in ('ignore', 'raise') else True
+    if errors == 'ignore':
```
```
try:
   conv = lib.maybe_convert........
except:
   if errors == 'raise':
        raise
   conv = orig
```

simpler no?

bashtage commented:
All addressed

bashtage commented:
@jreback It is green - or should be since all required tests passed.

jreback commented on pandas/tools/util.py:
```diff
@@ -48,3 +48,60 @@ def compose(*funcs):
+    if index is not None:
+        return pd.Series(conv, index=index, name=name)
+    else:
+        return conv
```
this will return an ndarray if you pass one in, indicate this in the Return Type above (and remove ref to box)

jreback commented:
@bashtage thanks for all of the work on this!
